### PR TITLE
evm: Transaction receipts

### PR DESCRIPTION
**Pull request by shohamc1**
## Summary

- Add transaction receipts
- `eth_getTransactionReceipt` RPC

Missing:
- `contractAddress` should be populated for contract creation transactions
- gas fields and associated calculations
- logs and logsBloom

### CODE REVIEW

**Review by DocteurPing — APPROVED** [review]
LGTM

File: lib/ain-grpc/src/rpc.rs [hunk 1]
```diff
@@ -368,4 +370,9 @@ impl MetachainRPCServer for MetachainRPCModule {
     fn gas_price(&self) -> Result<String> {
         Ok(format!("{:#x}", 0))
     }
+
+    fn get_receipt(&self, hash: H256) -> Result<ReceiptResult> {
+        let receipt = self.handler.receipt.get_receipt(hash).unwrap();
+        Ok(ReceiptResult::from(receipt))
+    }
 }
```
Comment (Jouzo) [inline lib/ain-grpc/src/rpc.rs:378]:
```suggestion
    fn get_receipt(&self, hash: H256) -> Result<Option<ReceiptResult>> {
        self.handler
            .receipt
            .get_receipt(hash)
            .map_or(Ok(None), |receipt| Ok(Some(ReceiptResult::from(receipt))))
    }
```
Should return None when not found

File: lib/ain-grpc/src/rpc.rs [hunk 2]
```diff
@@ -105,6 +106,9 @@ pub trait MetachainRPC {
 
     #[method(name = "eth_gasPrice")]
     fn gas_price(&self) -> Result<String>;
+
+    #[method(name = "eth_getTransactionReceipt")]
+    fn get_receipt(&self, hash: H256) -> Result<ReceiptResult>;
```
Comment (Jouzo) [inline lib/ain-grpc/src/rpc.rs:111]:
```suggestion
    fn get_receipt(&self, hash: H256) -> Result<Option<ReceiptResult>>;
```

File: lib/ain-grpc/src/receipt.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    pub block_number: U256,
+    pub contract_address: H160,
+    pub cumulative_gas_used: U256,
+    pub effective_gas_price: U256,
+    pub from: H160,
+    pub gas_used: U256,
+    pub logs: Vec<Log>,
+    pub logs_bloom: String,
+    pub status: String,
+    pub to: H160,
```
Comment (Jouzo) [inline lib/ain-grpc/src/receipt.rs:18]:
```suggestion
    pub to: Option<H160>,
```

File: lib/ain-evm/src/receipt.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                    status_code: 0,
+                    used_gas: Default::default(),
+                    logs_bloom: Default::default(),
+                    logs: vec![],
+                }),
+                block_hash,
+                block_number,
+                tx_hash: tv2.hash(),
+                from: transaction.sender,
+                to: get_recipient(get_transaction_action(tv2.clone())),
```
Comment (Jouzo) [inline lib/ain-evm/src/receipt.rs:156]:
```suggestion
                to: transaction.to(),
```

File: lib/ain-evm/src/receipt.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+                    status_code: 1,
+                    used_gas: Default::default(),
+                    logs_bloom: Default::default(),
+                    logs: vec![],
+                }),
+                block_hash,
+                block_number,
+                tx_hash: tv2.hash(),
+                from: transaction.sender,
+                to: get_recipient(get_transaction_action(tv2.clone())),
```
Comment (Jouzo) [inline lib/ain-evm/src/receipt.rs:129]:
```suggestion
                to: transaction.to(),
```

File: lib/ain-evm/src/receipt.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        TransactionV2::EIP1559(t) => t.action,
+    }
+}
+
+fn get_recipient(action: TransactionAction) -> H160 {
+    match action {
+        TransactionAction::Create => H160::zero(),
+        TransactionAction::Call(t) => t,
+    }
+}
```
Comment (Jouzo) [inline lib/ain-evm/src/receipt.rs:61]: Available in SignedTx

File: lib/ain-evm/src/receipt.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    return H160::from(keccak(stream.as_raw()));
+}
+
+fn get_nonce(tx: TransactionV2) -> U256 {
+    match tx {
+        TransactionV2::Legacy(t) => t.nonce,
+        TransactionV2::EIP2930(t) => t.nonce,
+        TransactionV2::EIP1559(t) => t.nonce,
+    }
+}
```
Comment (Jouzo) [inline lib/ain-evm/src/receipt.rs:77]: Available in SignedTx

File: lib/ain-evm/src/receipt.rs [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+pub static RECEIPT_MAP_PATH: &str = "receipt_map.bin";
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Receipt {
+    pub tx_hash: H256,
+    pub receipt: ReceiptV3,
+    pub block_hash: H256,
+    pub block_number: U256,
+    pub from: H160,
+    pub to: H160,
```
Comment (Jouzo) [inline lib/ain-evm/src/receipt.rs:28]:
```suggestion
    pub to: Option<H160>,
```

**Review by Jouzo — APPROVED** [review]

**Pull request merged**